### ISSUE TITLE
 Using a multiselect product attribute with a custom source model in the adminhtml doesn't render selected value.

### ISSUE THREAD

**Issue by kanduvisla** [magento/magento2#7537]
### Preconditions
Magento 2.4-develop

### Issue Description
I've created a custom source model to use as a product attribute of the type 'multiselect'. The Source model extends `Magento\Eav\Model\Entity\Attribute\Source\AbstractSource` and it shows up just fine. Saving the entities also work: the value gets stored in the database.

However, when the product page is loaded in the adminhtml, the proper values or not re-selected. I've tracked this down to the JavaScript configuration used for the attribute. If I create a multiselect product attribute from the adminhtml, I see in the JSON that the value of the options are strings:

    {"value":"4","label":"Foo"},{"value":"5","label":"Bar"}

However, the values of my custom source model are rendered as integers:

    {"value":4,"label":"Foo"},{"value":5,"label":"Bar"}

I've debugged the JavaScript by setting breakpoints in the inspector and everything is parsed correctly. That is: the values from the database are passed through to Magento. The only difference is the `value` in the JavaScript object:

![screen shot 2016-11-23 at 17 30 27](https://cloud.githubusercontent.com/assets/346681/20570097/9a3425c0-b1a2-11e6-82f7-2101ab2f3033.png)

![screen shot 2016-11-23 at 17 30 40](https://cloud.githubusercontent.com/assets/346681/20570099/9d15ffac-b1a2-11e6-89c2-c171a95b9794.png)

So I'm guessing that when you use custom source models, the values are not passed as strings, causing Knockout to not set the proper values

# Steps to reproduce

## 1. Generate a custom EAV source model.

For example:

```php
use Magento\Customer\Api\GroupManagementInterface;

/**
 * Class CustomerGroup
 */
class CustomerGroup 
    extends \Magento\Eav\Model\Entity\Attribute\Source\AbstractSource
    implements \Magento\Framework\Option\ArrayInterface
{
    /**
     * @var GroupManagementInterface
     */
    protected $groupManagement;

    /**
     * CustomerGroup constructor.
     * @param GroupManagementInterface $groupManagement
     */
    public function __construct(
        GroupManagementInterface $groupManagement
    )
    {
        $this->groupManagement = $groupManagement;
    }

    /**
     * @return array
     */
    public function toArray()
    {
        $groups = [];

        foreach ($this->groupManagement->getLoggedInGroups() as $group) {
            $groups[$group->getId()] = $group->getCode();
        }

        return $groups;
    }

    /**
     * Options getter
     * @return array
     */
    final public function toOptionArray()
    {
        $arr = $this->toArray();
        $ret = [];

        foreach ($arr as $key => $value) {
            $ret[] = [
                'value' => $key,
                'label' => $value
            ];
        }

        return $ret;
    }

    /**
     * Get options in "key-value" format
     * @return array
     */
    public function toArray()
    {
        return [];
    }

    /**
     * @return array
     */
    public function getAllOptions()
    {
        return $this->toOptionArray();
    }
}
```

## 3. Add it to an attribute set and save a product with some values

- Check the database: values are stored.
- Reload the product detail page: options are not selected, although they are present in the JS configuration / Knockout

---

It could be that I'm missing a step here with custom source models, but this always works with `select`-elements, but `multiselect` works different.
### Actual Result:
 values of my custom source model are rendered as integers

### Expected Result:
 values of my custom should be rendered as a string
## Update:

I'm currently able to fix this by making sure that the `toOptionArray()`-method of my source model always returns a string:

    /**
     * Options getter
     * @return array
     */
    final public function toOptionArray()
    {
        $arr = $this->toArray();
        $ret = [];

        foreach ($arr as $key => $value) {
            $ret[] = [
                // Always return a string:
                'value' => (string) $key,
                'label' => $value
            ];
        }

        return $ret;
    }

However, in my opinion this issue can still be considered a bug, because for semantics it's very strange I have to force my ID's (which are integers) to strings, just to get the Knockout multiselect component working.

Besides that, I have no idea what other things might break from this 'fix', since source models are broadly accepted and it might throw exceptions when an integer is expected.

**Comment by guido7171** [comment]
Hi Kandusivla Giel
I had the same issue. Some multi select attributes weren't saved. 
But some were! Only the ones where attribute_code starts with 0 and characters a-z did.
Then i tried to create the attribute in de admin and the following message appeared.

> Please use only letters (a-z), numbers (0-9) or underscore (_) in this field, **and the first character should be a letter.**

It seems there is a special condition for naming attribute_codes.

**Comment by kanduvisla** [comment]
@guido7171 That's true, but my attributes are named according to those conditions.

**Comment by guido7171** [comment]
@kanduvisla just to be sure, we're both talking about table eav_attribute column attribute_code ?

**Comment by kanduvisla** [comment]
Yes. The name of the attribute is "Included Customer Groups" and the code is "included_customer_groups". It's a multiselect attribute that has it's source model set to above example.

**Comment by kanduvisla** [comment]
Is there any news or at least some confirmation on this ticket?

**Comment by kanduvisla** [comment]
I've added a little update to my original question on how I currently fixed this.

**Comment by magento-engcom-team** [comment]
@kanduvisla, thank you for your report.
We've created internal ticket(s) MAGETWO-84479 to track progress on the issue.

**Comment by mukdam** [comment]
The same issue create a custom attribute programmatically on product multi-select .
Values saved in db but not showing on form and multiselect.js also showing values but not selected.

**Comment by mukdam** [comment]
I tried and Got success and not a Magento bug.ASAP I will update the example for how can create multi-select with custom source in product

**Comment by mallikaverma** [comment]
I am working on this at #dmcdindia1

**Comment by magento-engcom-team** [comment]
@engcom-Alfa Thank you for verifying the issue.

Unfortunately, not enough information was provided to acknowledge ticket. Please consider adding the following:
- [ ] Update issue content to follow format required by [Issue Reporting Guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines)

Once all required information is added, please add label `"Issue: Confirmed"` again.
Thanks!

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Alfa 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-30057` were created

**Issue Available**: @engcom-Alfa, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by engcom-Alfa** [comment]
Reproduced on the fresh 2.4-develop instance

**Actual Result:**
![screen](https://user-images.githubusercontent.com/51679138/71408773-02851b80-2648-11ea-870a-8355da282b39.png)

**Comment by engcom-Kilo** [comment]
Hello, @kanduvisla 
Thanks for your report.
I'm not able to reproduce this issue on latest 2.4-develop
![MultiselectCustomSourceModel](https://user-images.githubusercontent.com/60192090/103529911-2b610300-4e8f-11eb-8399-782baef13161.gif)
Please feel free to add more details and I'll reopen this issue.

**Comment by magento-engcom-team** [comment]
Hi @kanduvisla. 

Thank you for your report and collaboration!

The related internal Jira ticket `MC-30057` was closed as `Fixed`. 

The fix will be available with the upcoming `2.4.3` release.
